File: backend-node/src/evidence/evidence.controller.ts
import {
  BadRequestException,
  Body,
  Controller,
  Delete,
  Get,
  Param,
  Patch,
  Post,
  Query,
  UploadedFile,
  UseInterceptors,
} from '@nestjs/common'
import { FileInterceptor } from '@nestjs/platform-express'
import { Roles } from '../auth/http-auth.decorators'
import { JavaClientService } from '../common/java-client/java-client.service'
import { BatchReviewEvidenceDto } from './dto/batch-review-evidence.dto'
import { BatchTagEvidenceDto } from './dto/batch-tag-evidence.dto'
import { EvidenceCodeParamDto } from './dto/evidence-code.dto'
import { SearchEvidenceDto } from './dto/search-evidence.dto'
import { UpdateEvidenceMetadataDto } from './dto/update-evidence-metadata.dto'

interface UploadedEvidenceFile {
  buffer: Buffer
  originalname: string
  mimetype: string
}

@Controller('evidence')
export class EvidenceController {
  constructor(private readonly javaClient: JavaClientService) {}

  @Get()
  list(
    @Query('taskCode') taskCode?: string,
    @Query('alarmEventCode') alarmEventCode?: string,
  ) {
    if (!taskCode?.trim() && !alarmEventCode?.trim()) {
      throw new BadRequestException(
        '请至少提供 taskCode 或 alarmEventCode',
      )
    }
    const parameters = new URLSearchParams()
    if (taskCode?.trim()) {
      parameters.set('taskCode', taskCode.trim())
    }
    if (alarmEventCode?.trim()) {
      parameters.set('alarmEventCode', alarmEventCode.trim())
    }
    return this.javaClient.get(`/evidence?${parameters.toString()}`)
  }

  @Get('tags')
  tags() {
    return this.javaClient.get('/evidence/tags')
  }

  @Get('search')
  search(@Query() query: SearchEvidenceDto) {
    const parameters = new URLSearchParams()
    parameters.set('page', String(query.page ?? 0))
    parameters.set('size', String(query.size ?? 20))
    if (query.taskCode?.trim()) {
      parameters.set('taskCode', query.taskCode.trim())
    }
    if (query.alarmEventCode?.trim()) {
      parameters.set('alarmEventCode', query.alarmEventCode.trim())
    }
    if (query.deviceCode?.trim()) {
      parameters.set('deviceCode', query.deviceCode.trim())
    }
    if (query.assetType) {
      parameters.set('assetType', query.assetType)
    }
    if (query.sourceType) {
      parameters.set('sourceType', query.sourceType)
    }
    if (query.reviewStatus) {
      parameters.set('reviewStatus', query.reviewStatus)
    }
    if (query.startTime) {
      parameters.set('startTime', query.startTime)
    }
    if (query.endTime) {
      parameters.set('endTime', query.endTime)
    }
    if (query.keyword?.trim()) {
      parameters.set('keyword', query.keyword.trim())
    }
    if (query.includeDeleted != null) {
      parameters.set('includeDeleted', String(query.includeDeleted))
    }
    return this.javaClient.get(`/evidence/search?${parameters.toString()}`)
  }

  @Get(':evidenceCode')
  detail(@Param() params: EvidenceCodeParamDto) {
    return this.javaClient.get(
      `/evidence/${encodeURIComponent(params.evidenceCode)}`,
    )
  }

  @Post()
  @Roles('ADMIN', 'OPERATOR')
  @UseInterceptors(
    FileInterceptor('file', {
      limits: { fileSize: 20 * 1024 * 1024 },
    }),
  )
  upload(
    @UploadedFile() file?: UploadedEvidenceFile,
    @Body('taskCode') taskCode?: string,
    @Body('alarmEventCode') alarmEventCode?: string,
    @Body('deviceCode') deviceCode?: string,
  ) {
    if (!file) {
      throw new BadRequestException('请选择需要上传的证据文件')
    }
    return this.javaClient.postMultipart('/evidence', file, {
      taskCode,
      alarmEventCode,
      deviceCode,
    })
  }

  @Patch(':evidenceCode/metadata')
  @Roles('ADMIN', 'OPERATOR')
  updateMetadata(
    @Param() params: EvidenceCodeParamDto,
    @Body() body: UpdateEvidenceMetadataDto,
  ) {
    return this.javaClient.patch(
      `/evidence/${encodeURIComponent(params.evidenceCode)}/metadata`,
      body,
    )
  }

  @Post('batch/review')
  @Roles('ADMIN', 'OPERATOR')
  batchReview(@Body() body: BatchReviewEvidenceDto) {
    return this.javaClient.post('/evidence/batch/review', body)
  }

  @Post('batch/tags')
  @Roles('ADMIN', 'OPERATOR')
  batchTags(@Body() body: BatchTagEvidenceDto) {
    return this.javaClient.post('/evidence/batch/tags', body)
  }

  @Post(':evidenceCode/preview-url')
  @Roles('ADMIN', 'OPERATOR')
  previewUrl(@Param() params: EvidenceCodeParamDto) {
    return this.javaClient.post(
      `/evidence/${encodeURIComponent(params.evidenceCode)}/preview-url`,
      {},
    )
  }

  @Post(':evidenceCode/download-url')
  @Roles('ADMIN', 'OPERATOR')
  downloadUrl(@Param() params: EvidenceCodeParamDto) {
    return this.javaClient.post(
      `/evidence/${encodeURIComponent(params.evidenceCode)}/download-url`,
      {},
    )
  }

  @Delete(':evidenceCode')
  @Roles('ADMIN', 'OPERATOR')
  remove(@Param() params: EvidenceCodeParamDto) {
    return this.javaClient.delete(
      `/evidence/${encodeURIComponent(params.evidenceCode)}`,
    )
  }

  @Post(':evidenceCode/restore')
  @Roles('ADMIN', 'OPERATOR')
  restore(@Param() params: EvidenceCodeParamDto) {
    return this.javaClient.post(
      `/evidence/${encodeURIComponent(params.evidenceCode)}/restore`,
      {},
    )
  }
}

File: backend-java/src/main/java/com/skytrace/backend/evidence/EvidenceController.java
package com.skytrace.backend.evidence;

import com.skytrace.backend.common.ApiResponse;
import com.skytrace.backend.evidence.dto.BatchReviewEvidenceRequest;
import com.skytrace.backend.evidence.dto.BatchTagEvidenceRequest;
import com.skytrace.backend.evidence.dto.CreateEvidenceArchiveJobRequest;
import com.skytrace.backend.evidence.dto.EvidenceAccessUrlResponse;
import com.skytrace.backend.evidence.dto.EvidenceArchiveAccessUrlResponse;
import com.skytrace.backend.evidence.dto.EvidenceArchiveJobResponse;
import com.skytrace.backend.evidence.dto.EvidenceAssetResponse;
import com.skytrace.backend.evidence.dto.EvidenceDetailResponse;
import com.skytrace.backend.evidence.dto.EvidencePageResponse;
import com.skytrace.backend.evidence.dto.EvidenceSearchRequest;
import com.skytrace.backend.evidence.dto.EvidenceTagResponse;
import com.skytrace.backend.evidence.dto.EvidenceUploadResponse;
import com.skytrace.backend.evidence.dto.UpdateEvidenceMetadataRequest;
import com.skytrace.backend.evidence.service.EvidenceAccessService;
import com.skytrace.backend.evidence.service.EvidenceArchiveService;
import com.skytrace.backend.evidence.service.EvidenceCommandService;
import com.skytrace.backend.evidence.service.EvidenceMetadataService;
import com.skytrace.backend.evidence.service.EvidenceQueryService;
import com.skytrace.backend.evidence.service.EvidenceTagService;
import jakarta.validation.Valid;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;

import java.time.Instant;
import java.util.List;

@RestController
@RequestMapping("/evidence")
@ConditionalOnProperty(name = "app.minio.enabled", havingValue = "true")
public class EvidenceController {

    private final EvidenceQueryService queryService;
    private final EvidenceCommandService commandService;
    private final EvidenceAccessService accessService;
    private final EvidenceArchiveService archiveService;
    private final EvidenceMetadataService metadataService;
    private final EvidenceTagService tagService;

    public EvidenceController(
            EvidenceQueryService queryService,
            EvidenceCommandService commandService,
            EvidenceAccessService accessService,
            EvidenceArchiveService archiveService,
            EvidenceMetadataService metadataService,
            EvidenceTagService tagService) {
        this.queryService = queryService;
        this.commandService = commandService;
        this.accessService = accessService;
        this.archiveService = archiveService;
        this.metadataService = metadataService;
        this.tagService = tagService;
    }

    @GetMapping
    public ApiResponse<List<EvidenceAssetResponse>> list(
            @RequestParam(value = "taskCode", required = false) String taskCode,
            @RequestParam(value = "alarmEventCode", required = false)
            String alarmEventCode) {
        return ApiResponse.ok(queryService.findLegacy(taskCode, alarmEventCode));
    }

    @GetMapping("/tags")
    public ApiResponse<List<EvidenceTagResponse>> tags() {
        return ApiResponse.ok(tagService.listAll());
    }

    @GetMapping("/search")
    public ApiResponse<EvidencePageResponse> search(
            @RequestParam(value = "page", required = false) Integer page,
            @RequestParam(value = "size", required = false) Integer size,
            @RequestParam(value = "taskCode", required = false) String taskCode,
            @RequestParam(value = "alarmEventCode", required = false)
            String alarmEventCode,
            @RequestParam(value = "deviceCode", required = false)
            String deviceCode,
            @RequestParam(value = "assetType", required = false)
            String assetType,
            @RequestParam(value = "sourceType", required = false)
            String sourceType,
            @RequestParam(value = "reviewStatus", required = false)
            String reviewStatus,
            @RequestParam(value = "startTime", required = false)
            @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
            Instant startTime,
            @RequestParam(value = "endTime", required = false)
            @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
            Instant endTime,
            @RequestParam(value = "keyword", required = false) String keyword,
            @RequestParam(value = "includeDeleted", required = false)
            Boolean includeDeleted) {
        return ApiResponse.ok(queryService.search(new EvidenceSearchRequest(
                page,
                size,
                taskCode,
                alarmEventCode,
                deviceCode,
                assetType,
                sourceType,
                reviewStatus,
                startTime,
                endTime,
                keyword,
                includeDeleted
        )));
    }

    @GetMapping("/{evidenceCode}")
    public ApiResponse<EvidenceDetailResponse> detail(
            @PathVariable String evidenceCode) {
        return ApiResponse.ok(queryService.detail(evidenceCode));
    }

    @PostMapping(consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    public ApiResponse<EvidenceUploadResponse> upload(
            @RequestParam("file") MultipartFile file,
            @RequestParam(value = "taskCode", required = false) String taskCode,
            @RequestParam(value = "alarmEventCode", required = false)
            String alarmEventCode,
            @RequestParam(value = "deviceCode", required = false)
            String deviceCode) {
        return ApiResponse.ok(commandService.upload(
                file,
                taskCode,
                alarmEventCode,
                deviceCode
        ));
    }

    @PatchMapping("/{evidenceCode}/metadata")
    public ApiResponse<Void> updateMetadata(
            @PathVariable String evidenceCode,
            @RequestBody UpdateEvidenceMetadataRequest request) {
        metadataService.updateMetadata(evidenceCode, request);
        return ApiResponse.ok(null);
    }

    @PostMapping("/batch/review")
    public ApiResponse<Void> batchReview(
            @RequestBody BatchReviewEvidenceRequest request) {
        metadataService.batchReview(request);
        return ApiResponse.ok(null);
    }

    @PostMapping("/batch/tags")
    public ApiResponse<Void> batchTags(
            @RequestBody BatchTagEvidenceRequest request) {
        metadataService.batchTags(request);
        return ApiResponse.ok(null);
    }

    @PostMapping("/{evidenceCode}/preview-url")
    public ApiResponse<EvidenceAccessUrlResponse> previewUrl(
            @PathVariable String evidenceCode) {
        return ApiResponse.ok(accessService.createPreviewUrl(evidenceCode));
    }

    @PostMapping("/{evidenceCode}/download-url")
    public ApiResponse<EvidenceAccessUrlResponse> downloadUrl(
            @PathVariable String evidenceCode) {
        return ApiResponse.ok(accessService.createDownloadUrl(evidenceCode));
    }

    @PostMapping("/archive-jobs")
    public ApiResponse<EvidenceArchiveJobResponse> createArchiveJob(
            @Valid @RequestBody CreateEvidenceArchiveJobRequest request) {
        return ApiResponse.ok(archiveService.createJob(request));
    }

    @GetMapping("/archive-jobs/{jobCode}")
    public ApiResponse<EvidenceArchiveJobResponse> archiveJob(
            @PathVariable String jobCode) {
        return ApiResponse.ok(archiveService.getJob(jobCode));
    }

    @PostMapping("/archive-jobs/{jobCode}/download-url")
    public ApiResponse<EvidenceArchiveAccessUrlResponse> archiveDownloadUrl(
            @PathVariable String jobCode) {
        return ApiResponse.ok(archiveService.createDownloadUrl(jobCode));
    }

    @PostMapping("/archive-jobs/{jobCode}/manifest-url")
    public ApiResponse<EvidenceArchiveAccessUrlResponse> archiveManifestUrl(
            @PathVariable String jobCode) {
        return ApiResponse.ok(archiveService.createManifestUrl(jobCode));
    }

    @DeleteMapping("/{evidenceCode}")
    public ApiResponse<Void> delete(@PathVariable String evidenceCode) {
        commandService.softDelete(evidenceCode);
        return ApiResponse.ok(null);
    }

    @PostMapping("/{evidenceCode}/restore")
    public ApiResponse<Void> restore(@PathVariable String evidenceCode) {
        commandService.restore(evidenceCode);
        return ApiResponse.ok(null);
    }
}
